### PR TITLE
Constructors, records, void returns, method descs.

### PR DESCRIPTION
**Pull request by aomarks**
- Support constructors (e.g. `function(new:HTMLElement, string)`).
- Support record types (e.g. `@param {{foo: bar}}`).
- Support `void` function return types.
- Include method `@returns` descriptions.

Resolves a number of comments in https://github.com/Polymer/polymer/pull/4928

### CODE REVIEW

File: src/test/goldens/polymer/lib/legacy/legacy-element-mixin.d.ts [hunk 1]
```diff
@@ -238,7 +238,7 @@ declare namespace Polymer {
      *  `node` on which to fire the event (HTMLElement, defaults to `this`).
      * @returns The new event that was fired.
      */
-    fire(type: string, detail?: any, options?: any): Event|null;
+    fire(type: string, detail?: any, options?: {bubbles: boolean|undefined, cancelable: boolean|undefined, composed: boolean|undefined}): Event|null;
```
Comment (TimvdLippe) [inline src/test/goldens/polymer/lib/legacy/legacy-element-mixin.d.ts]: Nit: |undefined can be an optional parameter

Comment (aomarks) [reply src/test/goldens/polymer/lib/legacy/legacy-element-mixin.d.ts]: Done.

File: src/closure-types.ts [hunk 2]
```diff
@@ -181,15 +184,61 @@ function convertUnion(
...
+      fieldType.members = fieldType.members.filter((member) => {
+        if (member.kind === 'name' && member.name === 'undefined') {
+          optional = true;
+          return false;
+        }
+        return true;
+      });
+
+      // No need for a union if we collapsed it to one member.
+      if (fieldType.members.length === 1) {
```
Comment (rictic) [inline src/closure-types.ts]: Isn't there code elsewhere that does this? Can it be unified with this code?

File: src/test/closure-types_test.ts [hunk 3]
```diff
@@ -111,6 +113,19 @@ suite('closureTypeToTypeScript', () => {
     check('!Function', 'Function');
   });
 
+  test('constructor', () => {
+    check('function(new:HTMLElement)', '{new(): HTMLElement}');
+    check(
+        'function(new:HTMLElement, string)', '{new(p0: string): HTMLElement}');
```
Comment (rictic) [inline src/test/closure-types_test.ts:119]: Wow that's weird closure syntax. TIL

**Review by rictic — APPROVED** [review]

File: src/closure-types.ts (hunk 2 above)
Comment (aomarks) [reply src/closure-types.ts]: Done, just calling simplify() now.

**Pull request merged**